### ISSUE TITLE
backward compatibility - need a way to find out when a certain API was added/modified/etc.

### ISSUE THREAD

**Issue by stas00** [pytorch/pytorch#60149]
## 📚 Documentation

Currently it's very difficult to know when an API has been modified, which is a requirement for backward compatibility. 

Let's take: https://pytorch.org/docs/stable/generated/torch.Tensor.add_.html which starting from 1.9.0 deprecates
```
UserWarning: This overload of add_ is deprecated:
        add_(Number alpha, Tensor other)
Consider using one of the following signatures instead:
        add_(Tensor other, *, Number alpha) (Triggered internally at  /pytorch/torch/csrc/utils/python_arg_parser.cpp:1025.)
  exp_avg_sq.mul_(beta2t).add_(1.0 - beta2t, update)
```

But I don't know whether it's safe to do `Tensor.add_(other, *, alpha=1)` because I don't know when this variation of API was added, so if I have to support say, `pytorch==1.2` will it still work or will I have to write a wrapper which will call a different version of `add_` depending on the pytorch version.

I can go to git and try to painfully hunt it down, which is additionally more difficult for raw Cpp functions, but surely there must be a better way.

Thank you!

**Comment by gchanan** [comment]
it would be good if our documentation listed when functions were added / modified, etc.

I would have probably done a binary search through the actual releases, once you have them all set up it's probably faster.

**Comment by stas00** [comment]
> I would have probably done a binary search through the actual releases, once you have them all set up it's probably faster.

Right,  documenting an easy - ideally automated process for doing this work faster would be helpful too.

I also realize that it might take a lot of work to figure out things that are already there, but perhaps a process can be developer for any new addition/modification of the function from now and onward.

**Comment by ezyang** [comment]
`native_functions.yaml` has been around for a long enough time and has stayed relatively stable, so I bet that we can retroactively compute these API changes by comparing JIT schemas. This would be a fun project.

**Comment by stas00** [comment]
Great!

Perhaps even in a simple variation of a grep, as proposed here:
https://github.com/pytorch/pytorch/issues/60153#issuecomment-863762810

**Comment by alanwaketan** [comment]
Taking this as a bootcamp task.

**Comment by alanwaketan** [comment]
I had an initial question regarding to the scope of this issue:
1. Are we trying to find an official way of adding API availability to the documentation like: [WKWebView](https://developer.apple.com/documentation/webkit/wkwebview)? It means we probably need to: 1) label all existing API, probably in native_functions.yaml, with categories: PyTorch 1.0 - 1.9, and TBD. 2) build a release tool that once a new version, e.g. 1.10, is released, it automatically changes all TBD to the new version. 3) build some other tools that will pick up those information and display them to the developer properly.
2. Or a subset of the above?
3. Or just a Python script that accepts a signature from native_functions.yaml, i.e., `add_.Tensor(Tensor(a!) self, Tensor other, *, Scalar alpha=1) -> Tensor(a!)`, and returns a version when it was shipped, i.e., 'pytorch-1.5.0' (probably through bisecting the releases)?

**Comment by ezyang** [comment]
There is definitely a number of different ways you could solve the underlying issue.

For a bootcamp task, I think a well scoped formulation is as follows: write a script that a pair of `native_functions.yaml` (you will source these pairs by checking out various release tags of PyTorch and grabbing their copy of `aten/src/ATen/native/native_functions.yaml`. Then, you will do a semantic diff between the YAMLs. To do this, it is easiest to repurpose the existing `native_functions.yaml` parser in `tools/codegen/model.py` (you can see it in use in `tools/codegen/gen.py`. You may need to do a little work to make the parser a little less strict when there are fields it doesn't understand.

A simple semantic diff algorithm to implement is:
1. Match up all functions that have exactly the same function name and overload name
2. Check if the JIT schema is the same or not (if it is not the same, that means BC-breaking change, or new attribute addition)

Run this between, e.g., 1.9 and 1.8, and show that it provides some reasonable results.

**Comment by alanwaketan** [comment]
> There is definitely a number of different ways you could solve the underlying issue.
> 
> For a bootcamp task, I think a well scoped formulation is as follows: write a script that a pair of `native_functions.yaml` (you will source these pairs by checking out various release tags of PyTorch and grabbing their copy of `aten/src/ATen/native/native_functions.yaml`. Then, you will do a semantic diff between the YAMLs. To do this, it is easiest to repurpose the existing `native_functions.yaml` parser in `tools/codegen/model.py` (you can see it in use in `tools/codegen/gen.py`. You may need to do a little work to make the parser a little less strict when there are fields it doesn't understand.

It looks like model.py is not quite backward compatible.😂 I will upload a WIP patch that just disables all the assertion manually. Then we can discuss how we can introduce a loose mode to the parser.

**Comment by ezyang** [comment]
yes, please do that

**Comment by stas00** [comment]
Any chance this can be revived?

Had to look for when `torch.linalg.norm` was added to properly support older pytorch in HF Transformers, and it's very slow and painful process.

I wonder if at least from now and forward each newly added function requires to say:

```
Added in torch 1.10
```

It doesn't help that one can't quickly switch the version of the docs on a given function which could have been another way to relatively quickly find out the version where it disappears, e.g. going from 

https://pytorch.org/docs/stable/generated/torch.linalg.norm.html
https://pytorch.org/docs/1.9.0/generated/torch.linalg.norm.html
https://pytorch.org/docs/1.8.0/generated/torch.linalg.norm.html

via switching the version number in the left menu. But this is not possible either, it takes me to the main page of the release :(

Thank you.

**Comment by zou3519** [comment]
@alanwaketan are you still working on this? If not we should release it so that someone else can pick it up

**Comment by alanwaketan** [comment]
> @alanwaketan are you still working on this? If not we should release it so that someone else can pick it up

Nope, I don't think so. Feel free to take the task.

**Comment by Lezcano** [comment]
When thinking about @ezyang's proposal, I realised that `test/backward_compatibility/check_backward_compatibility.py` pretty much flags when there's a change in `native_functions.yaml`. I have not looked into the logic there, but perhaps that logic can be reused / repurposed for this.

**Comment by ezyang** [comment]
Check BC won't help, as it only flags BC incompatible changes. 

BTW, if someone wants to keep going on the mechanism I suggested, https://github.com/pytorch/pytorch/pull/61374#issuecomment-877203584 is where I summarized next steps.

**Comment by mruberry** [comment]
fyi our updated BC/FC guide directs developers to use the "versionchanged" Sphinx directive: https://github.com/pytorch/pytorch/wiki/PyTorch's-Python-Frontend-Backward-and-Forward-Compatibility-Policy.

**Comment by albanD** [comment]
Bumping to triage meeting for visibility

**Comment by malfet** [comment]
Very similar discussion on the forums: https://dev-discuss.pytorch.org/t/how-to-keep-up-with-updates-of-operator-signatures/834

**Comment by janeyx99** [comment]
Some action items for devs:
1. use the versionchanged directive when modifying schemas in docs
2. be explicit in release notes to increase searchability of an API

**Comment by stas00** [comment]
Almost 3 years passed and the user still has no way of knowing when this or that API was added or modified.

e.g. here is the latest example: a user always has to ask: https://github.com/pytorch/pytorch/pull/121364#issuecomment-2023625314

**Comment by albanD** [comment]
I think it is a good expectation to force use of the following in our doc:
- `.. versionadded::`
- `.. versionchanged::`
- `.. deprecated::`

The main question is how do we enforce the use of these? The main options are:
- Soft: we broadly asks all our devs to please do this properly
- Medium: we add expectations for all reviewers that they should ask for these directives when applicable
- Hard: find a way to check this in CI. Not sure how we can do this.

WDYT?

**Comment by msaroufim** [comment]
At a high level I think does require some initial heroics to document every single public function in PyTorch and once most of them are documented then we can totally enforce either the medium or hard workflow.

At least for CI I'd expect changes to function arguments or return type to be sufficient checks

**Comment by albanD** [comment]
> At least for CI I'd expect changes to function arguments or return type to be sufficient checks

The bc-linter is detecting such change in the code. But how do we enforce that the doc is appropriately updated?

**Comment by albanD** [comment]
cc @kit1980 can the bc-linter detect newly added APIs?
